splinterd/src/config/builder.rs

```diff
@@ -131,11 +154,29 @@ impl ConfigBuilder {
                 None => None,
             })
             .ok_or_else(|| ConfigError::MissingValue("network endpoints".to_string()))?;
+        let state_dir = self
+            .partial_configs
+            .iter()
+            .find_map(|p| match p.state_dir() {
+                Some(v) => Some((v, p.source())),
+                None => None,
+            })
+            .ok_or_else(|| ConfigError::MissingValue("state directory".to_string()))?;
+
         let node_id = self
             .partial_configs
             .iter()
             .find_map(|p| match p.node_id() {
-                Some(v) => Some((v, p.source())),
+                Some(v) => {
+                    if p.source() == ConfigSource::Default {
```

Comment:
If the node id file exists, it must always match the configuration or we should produce a fatal error.